### DISCUSSION TITLE
How to debug .NET Runtime tests inside Docker?

### DISCUSSION THREAD

**Discussion by Maximys** [dotnet/runtime#86943] [Q&A]
Hello everybody!
Can anybody explain me how to build .NET Runtime and debug tests inside Docker? I trying next command:
`docker run -v C:\runtime\:/runtime -w /runtime mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-18.04 ./build.sh`
but result is:
```
/runtime/eng/common/tools.sh: line 501: /runtime//.dotnet/dotnet: No such file or directory
Build failed with exit code 127. Check errors above.
```

**Comment by Clockwork-Muse** [comment]
You're going to have a much better time if you clone **inside** WSL and open the devcontainer (or clone inside a volume and then open the devcontainer).  For one thing, there are performance penalties for mounting the Windows filesystem inside of Linux (as you're doing with the `-v` volume mount).  For another, it will also enable you to get interactive access to the code.

**Reply by Maximys** [reply]
@Clockwork-Muse , thanks. Can you provide more detailed info about that?

**Comment by am11** [comment]
> ```
> /runtime/eng/common/tools.sh: line 501: /runtime//.dotnet/dotnet: No such file or directory
> Build failed with exit code 127. Check errors above.
> ```

This error means `runtime/.dotnet/dotnet` exists but it is not a linux ELF executable. Try deleting `artifacts` and `.dotnet` directories before running that docker command.

>  and debug tests inside Docker?

I use privileged and interactive mode when debugging in container is needed:
 `docker run --privileged -v$(pwd):/runtime -w /runtime --it $CID bash`
and install `lldb` inside the container. When running the .NET executable under `lldb`, remember to disable ASLR (gdb does this automatically but lldb does not):

```gdb
$ lldb -- /runtime/artifacts/<path/to/some_executable> <args of that executable>
(lldb) settings set target.disable-aslr false
# now you can set the breakpoints and `run` etc.
```
or start lldb as `lldb -o "settings set target.disable-aslr false" <rest of the command>` or put disable-aslr command in your `~/.lldbinit` file.

For more info, see https://github.com/dotnet/runtime/tree/main/docs/workflow/debugging; some of the info is missing in our docs, so feel free to send a PR based on what you experienced setting it up. Good luck!
